### PR TITLE
 Add SpecifyArgSeparatorFixer to make sure http_build_query is properly used

### LINKED ISSUE #17: Make sure http_build_query is used with third param
`arg_separator.output` could be changed to anything in PHP settings (eg. `&arg;`), which then unexpectedly breaks URLs assembled using `http_build_query()`.

To avoid this [unexpected errors](https://github.com/lmc-eu/matej-client-php/pull/54), we should make sure the [third parameter](http://php.net/manual/en/function.http-build-query.php) (arg separator) is always used.

### PR DESCRIPTION
**Pull request by OndraM**
Fixes #17 .

To be honest, this is the first Fixer I've ever written. So any suggestions especially regarding usage of php-cs-fixer internals (the tokens stuff etc.) are welcome. Also I may overlook some simpler way how to accomplish this, so I'm open for suggestions.

Cc @TomasVotruba - I would be grateful to hear some feedback on the Fixer from you as well :heart: :-)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -10,20 +10,45 @@
...
         }
     ],
+    "autoload": {
+        "psr-4": {
+            "Lmc\\CodingStandard\\": "src/"
+        }
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "Lmc\\CodingStandard\\": "tests/"
```
Comment (TomasVotruba) [inline composer.json]: Why not convetion `"Lmc\\CodingStandard\\Tests\\": "tests/"`?

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            [
+                new CodeSample("<?php\n\$queryString = http_build_query(['foo' => 'bar', 'baz' => 'bat']);"),
+            ],
+            'Function `http_build_query()` uses `arg_separator.output` ini directive as default argument separator, '
+            . 'however when its default value "&" is changed, query string assembled by the method will be '
+            . 'unexpectedly invalid. This Fixer forces you to not rely on ini settings and rather define '
+            . '`$arg_separator` in third parameter of the function.',
+            null,
+            null,
+            'Risky when other than default "&" argument separator should be used in query strings.'
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: I'd expect this line rather (even) at `isRisky()` method

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        foreach ($tokens as $index => $token) {
+            if ($token->isGivenKind(T_STRING) && $token->getContent() === 'http_build_query') {
+                $previousIndex = $tokens->getPrevMeaningfulToken($index);
+
+                // Function imports and function definitions should not be affected
+                if ($previousIndex !== null
+                    && ($tokens[$previousIndex]->isGivenKind(CT::T_FUNCTION_IMPORT)
+                        || $tokens[$previousIndex]->isGivenKind(T_FUNCTION))) {
+                    return;
+                }
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: I'd add all this to "shouldSkip()" method, it's not relevant to fixer process at all

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+                $this->fixFunction($tokens, $index);
+            }
+        }
+    }
+
+    public function getName(): string
+    {
+        return mb_strtolower(self::class);
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Why did you choose this new approach?

I'd use one of standards: php-cs-fixer or ECS. How could you copy this class to `easy-coding-standard.yml` from CLI output to `skip` it?

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    {
+        return true;
+    }
+
+    private function fixFunction(Tokens $tokens, int $functionIndex): void
+    {
+        $startBraceIndex = $tokens->getNextTokenOfKind($functionIndex, ['(']);
+        if ($startBraceIndex === null) {
+            return;
+        }
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Why is this here? Should be above the fix funciton

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+                continue;
+            }
+
+            if ($token->equals(',')) {
+                $commaIndexes[] = $index;
+
+                continue;
+            }
+        }
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Might be helpful:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.11/src/Tokenizer/Analyzer/ArgumentsAnalyzer.php

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        // When third parameter is already present, make sure it is not null or override its value
+        if ($paramsCount >= 3) {
+            $thirdParamTokenIndex = $tokens->getNextMeaningfulToken($commaIndexes[1]);
+            if ($thirdParamTokenIndex === null) {
+                return;
+            }
+            $thirdParamToken = $tokens[$thirdParamTokenIndex];
+
+            if ($thirdParamToken->isGivenKind(T_STRING) && $thirdParamToken->getContent() === 'null') {
+                $tokens->offsetSet($thirdParamTokenIndex, $ampToken);
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: This section very difficult to read. What does it do?

I'd expect it to be bellow, with other "modifying" code

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        $tokensToInsert = [];
+
+        // Add second parameter if not defined
+        if ($paramsCount === 1) {
+            $tokensToInsert[] = new Token(',');
+            $tokensToInsert[] = new Token([T_WHITESPACE, ' ']);
+            $tokensToInsert[] = new Token([T_STRING, 'null']);
+        }
+
+        // Add third parameter (arg separator)
```
Comment (TomasVotruba) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Real code might be also helfpul while reading this, e.g.:

```php
// Add third parameter (arg separator): ", &")
```

I always add it because it's much more readable than custom Tokens

File: tests/Fixer/SpecifyArgSeparatorFixerTest.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    public function shouldFixCode(string $inputFile, string $expectedOutputFile): void
+    {
+        $fixer = new SpecifyArgSeparatorFixer();
+        $fileInfo = new \SplFileInfo(__DIR__ . '/Fixtures/' . $inputFile);
+        $tokens = Tokens::fromCode(file_get_contents($fileInfo->getRealPath()));
+
+        $fixer->fix($fileInfo, $tokens);
+
+        $this->assertSame(
+            file_get_contents(__DIR__ . '/Fixtures/' . $expectedOutputFile),
```
Comment (TomasVotruba) [inline tests/Fixer/SpecifyArgSeparatorFixerTest.php]: `asssertStringEqualsFile()` or how its called

File: tests/Fixer/SpecifyArgSeparatorFixerTest.php [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            $tokens->generateCode()
+        );
+    }
+
+    public function provideFiles()
+    {
+        return [
+            'Correct file should not be changed' => ['Correct.txt', 'Correct.txt'],
+            'Wrong file should be fixed' => ['Wrong.txt', 'Fixed.txt'],
+        ];
```
Comment (TomasVotruba) [inline tests/Fixer/SpecifyArgSeparatorFixerTest.php]: While using dataProvider, it makes sense to use one file per case.
Imagine you have bug in Wrong.txt on line 25 - which one of these cases it?

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 6 above)
Comment (TomasVotruba) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: I'd be happy to check after that, there is lot of outsourcable code now

File: composer.json (hunk 1 above)
Comment (OndraM) [reply composer.json]: We use the same namespace for Class and its corresponding test.

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 4 above)
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: "The name must be all lowercase and without any spaces." - this is in phpDoc. 
But I saw you use class name. So I combined them both :-D. What is the best solution? Where is this name being used?

ECS is depending just on the class name FQCN, or it somehow works with this?

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 5 above)
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: Well I don't know why is this condition even here. Probably some forgotten code, I will remove it :).

File: tests/Fixer/SpecifyArgSeparatorFixerTest.php (hunk 10 above)
Comment (OndraM) [reply tests/Fixer/SpecifyArgSeparatorFixerTest.php]: @TomasVotruba True, but there is overhead and is also harder to see what is covered and what not. Maybe rather write inline PHP for each case (as some fixer tests do).

You will see which line of code is broken by phpunit's diff, right?

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 5 above)
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: `getNextTokenOfKind()` can by annotation return null as well. So to adhere type safety, this case must be somehow handled.

File: .travis.yml [hunk 11]
```diff
@@ -1,11 +1,17 @@
...
 php:
+  - 7.1
   - 7.2
 
 install:
   - travis_retry composer self-update && composer --version
   - travis_retry composer update --no-interaction
 
 script:
-  - composer all
```
Comment (MortalFlesh) [inline .travis.yml]: why not to encapsulate those commands into `all` script?

File: composer.json [hunk 12]
```diff
@@ -10,20 +10,45 @@
...
+        }
+    },
     "require": {
         "symplify/easy-coding-standard": "^4.0.2"
     },
     "require-dev": {
         "j13k/yaml-lint": "^1.1",
-        "mhujer/yaml-sort-checker": "^1.2"
+        "mhujer/yaml-sort-checker": "^1.2",
```
Comment (MortalFlesh) [inline composer.json]: Did you consider using `lmc/coding-standards` here as well?

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    {
+        return true;
+    }
+
+    public function fix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        foreach ($tokens as $index => $token) {
+            if ($token->isGivenKind(T_STRING) && $token->getContent() === 'http_build_query') {
+                if ($this->isNotFunctionCall($tokens, $index)) {
+                    return;
```
Comment (MortalFlesh) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Are you sure with `return`? IMHO this should be just `continue`, or am I wrong?

Imagine code
```php
function foo($args) {
    $this->addCallback('http_build_query'); // this is not a function call
    return http_build_query($args);  // this is a function call, but isn't it skipped due previous `return`?
}
```

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+    private function fixFunction(Tokens $tokens, int $functionIndex): void
+    {
+        $startBraceIndex = $tokens->getNextTokenOfKind($functionIndex, ['(']);
+        if ($startBraceIndex === null) {
+            return;
+        }
+
+        $endBraceIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $startBraceIndex);
+        $ampToken = new Token([T_STRING, "'&'"]);
```
Comment (MortalFlesh) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Well, would this `Token` find usage of `http_build_query(..., "&");`? You explicitely awaits it is in single quotes.

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+        $commaIndexes = [];
+
+        for ($index = $startBraceIndex + 1; $index < $endBraceIndex; ++$index) {
```
Comment (MortalFlesh) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: Just saying..  but this looks very confusing to me (`+ 1`, `++$index`,...).

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+            $token = $tokens[$index];
+
+            // Parameter surrounded by brace, skip to its end
+            if ($token->equals('(')) {
```
Comment (MortalFlesh) [inline src/Fixer/SpecifyArgSeparatorFixer.php]: mm.. I guess following code is a nightmare for Fixer author, but have you considered something like this?
```php
http_build_query([], null, ('&'));
```

File: composer.json (hunk 12 above)
Comment (OndraM) [reply composer.json]: Why :)? This is the package itself :upside_down_face:

File: .travis.yml (hunk 11 above)
Comment (OndraM) [reply .travis.yml]: I need phpunit with specific settings for UI (to generate clover). But I may do something like "all-ci" or so.

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 13 above)
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: Nice catch, fixed. :+1:

**Comment by OndraM** [comment]
Changes done & refactored to use ArgumentsAnalyzer. cc @TomasVotruba @MortalFlesh et al. :pray:

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 14 above)
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: Good point. Priority is now set to -1, ie. it will run after SingleQuoteFixer.

File: tests/Fixer/SpecifyArgSeparatorFixerTest.php (hunk 10 above)
Comment (TomasVotruba) [reply tests/Fixer/SpecifyArgSeparatorFixerTest.php]: Well, how do you run broken code only?

**Comment by TomasVotruba** [comment]
In which commit did you add the `ArgumentAnalyzer`? The commit messages are not very useful :D

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        $previousIndex = $tokens->getPrevMeaningfulToken($index);
+
+        return $previousIndex !== null
+            && ($tokens[$previousIndex]->isGivenKind(CT::T_FUNCTION_IMPORT)
+                || $tokens[$previousIndex]->isGivenKind(T_FUNCTION));
+    }
+
+    private function getThirdArgumentInfo(
+        Tokens $tokens,
+        $openParenthesisIndex,
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: How can this pass the CI?

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        $closeParenthesisIndex
+    ): ?TypeAnalysis {
+        $argumentsAnalyzer = new ArgumentsAnalyzer();
+
+        $arguments = $argumentsAnalyzer->getArguments($tokens, $openParenthesisIndex, $closeParenthesisIndex);
+        $argumentIndex = array_slice($arguments, 2, 1, true);
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: So is this `int` (index/position) or some array/object?

**Comment by OndraM** [comment]
@TomasVotruba 
> In which commit did you add the ArgumentAnalyzer? The commit messages are not very useful :(

You can easily go to Files changed > Changes from: > Select "Since last review" or something like this (it is selected by default) to see what has changed.

We don't write commit new messages when you only adjust things in response to code-review feedback, because commit messages should be about behavior change in the code, to make the history help with long term maintainability of the code (it helps a lot when using git blame etc.). There already is commit about adding the fixer, so unless any new commit in this PR will add/change some other behavior, the changes are just associated to the original commit via fixup and will be rebased  and squashed before merge.

Interesting article about this topic: https://arialdomartini.wordpress.com/2012/09/03/pre-emptive-commit-comments/ (especially "Tell me what the software does (not what you did)" part) and also of course https://filip-prochazka.com/blog/git-fixup about fixup commits.

File: src/Fixer/SpecifyArgSeparatorFixer.php [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        if (!empty($tokensToInsert)) {
+            $beforeCloseParenthesisIndex = $tokens->getPrevNonWhitespace($closeParenthesisIndex);
+            $tokens->insertAt($beforeCloseParenthesisIndex + 1, $tokensToInsert);
+        }
+    }
+
+    /**
+     * Return false if this is function import or function definition (and not function call)
+     */
+    private function isNotFunctionCall(Tokens $tokens, int $index): bool
```
[the remark this answers is not in the document]
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: Neither am I, but this function in fact doesn't prove it is function call when you just negate the logic - there are IMHO many edge cases. What it can is just to detect the main scenarios when it is not a function call :). So from my POV, the other name would be inaccurate and misleading :).
🤷

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: :scream:

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: The content is `[startIndex => endIndex]`. IMO argument is still misleading - it contains just the indexes, not the argument itself 🤔

**Comment by TomasVotruba** [comment]
It's just uncommon and confusing in open source, that's all.

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (MortalFlesh) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: Well.. 🤔 
Ok in this case I think it might be ok..

File: tests/Fixer/SpecifyArgSeparatorFixerTest.php (hunk 10 above)
Comment (MortalFlesh) [reply tests/Fixer/SpecifyArgSeparatorFixerTest.php]: You dont (I guess).. but why would you run just part of tests?
If you are developing something new, you can run single test case or I think even one perticular provider value test case in PHPStorm..

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: @TomasVotruba It seems phpdoc_add_missing_param_annotation (aka PhpdocAddMissingParamAnnotationFixer) does not work as I expected :-(. It only works if there is already at least something in phpdoc.

Do you know some check which takes care of this?

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: @MortalFlesh I thought about it again and changed it in the end... The objection still applies, but maybe its less harm like this :man_shrugging: .

File: src/Fixer/SpecifyArgSeparatorFixer.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (OndraM) [reply src/Fixer/SpecifyArgSeparatorFixer.php]: Oh, it seems new fixer for this will make it to php-cs-fixer 2.12 or so: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3640 :tada:

File: composer.json (hunk 1 above)
Comment (TomasVotruba) [reply composer.json]: I know :D but why not convention?

File: tests/Fixer/SpecifyArgSeparatorFixerTest.php (hunk 10 above)
Comment (TomasVotruba) [reply tests/Fixer/SpecifyArgSeparatorFixerTest.php]: I do that too, but how to run specific one line in this single file?

File: composer.json (hunk 1 above)
Comment (OndraM) [reply composer.json]: AFAIK both of the approaches are widely used, so I don't consider any of them as "general convention".

And why it should be in standalone namespace in the first place? I don't see a reason for this.

The same namespace means you don't need to import the class under test namespace, also it makes sense to me to have class and its tests in the same "module" (namespace), because they belongs to each another.

But its still more about personal preference...

Comment (TomasVotruba) [reply composer.json]: Similar unconvention as putting src and tests in the same directory, causes in php cs fixer only troubles. 80 % of php-scoper setup took me only this to solve.

It's really painful to work with such packages, where maintainer don't see behing the package common use.

Comment (OndraM) [reply composer.json]: Sorry, I don't understand :). This is the way we use it for years and never heard of single issue. And it is not any "hack",  just usual setting of Composer autoloading. But yeah, I can imagine this could cause trouble when developing some low-level stuff.

BTW `src/` and `tests/` in the same (root) directory are used almost everywhere except packages which are then splited to sub-packages (like Symfony). I find most inconvenient when source is not in standalone directory and mixed with `Tests` directory on the same level. Setting code-coverage or phpcs etc. blacklist/whitelist is nightmare in this case.

Comment (TomasVotruba) [reply composer.json]: Again, open source code is very different to closed one.

Comment (OndraM) [reply composer.json]: True, but I don't think the difference is in the way autoloading is set up :). But this seems like a topic to pub/meetup, so let's don't forget it next time we meet :-D . I'd like to hear about the stuff you deal with :).

**Comment by OndraM** [comment]
Merged, thanks @TomasVotruba @MortalFlesh a lot for your great feedback! :1st_place_medal:
